Add API function RunEnergyPlus that uses return codes instead of std::exit()

michael-okeefe opened the pull request:
Pull request overview
---------------------
**Summary**: This pull request adds an additional library interface (int RunEnergyPlus(...)) to EnergyPlus that passes return codes for success/failure back to the caller and also modifies streams and how streams are handled such that stdin and stdout are not accidentally closed or put in an errored state. The previous interface (void EnergyPlusPgm(...)) that calls std::exit() is retained for backwards compatibility. Note: this pull request contains no changes to the ObjexxFCL library. The behavior that was seen where stdin/stdout/stderr (cin/cout/cerr) was reported separately and a fix is being considered for the next revision of ObjexxFCL that will prevent stdin/stdout/stderr from going into an errored state. This PR contains a workaround that checks for whether cin/cout/cerr are good and, if not, clears their error bits which effectively resolves the issue.

-----------------

A new feature proposal for this effort is available: ["Add C-API for EnergyPlus DLL and API Method that uses Return Codes instead of std::exit()"](https://github.com/NREL/EnergyPlus/blob/capi_pull_request/design/FY2018/NFP-capi-and-return-codes.md)

-----------------

In recent work with using EnergyPlus as a library, EnergyPlus's default behavior to call std::exit has proved difficult to work with. As a clear example of what is not working correctly, please see file `src/EnergyPlus/test_ep_as_library.cc`.  Add the following line to the existing version of the code after the call to EnergyPlusPgm(...):

`std::cout << "THIS LINE SHOULD DISPLAY." << std::endl;`

If you run the modified `test_ep_as_library.cc` against the current EnergyPlus, the line is not printed out for two reasons:

1. The call to EnergyPlusPgm(...) (indirectly) calls std::exit(...) which exits the program (and anything calling EnergyPlus in the same process) and thus your code never returns from EnergyPlusPgm to do any follow-on tasks
2. Even if you modify EnergyPlusPgm(...) to not call std::exit(...), you will be able to return from the simulation but you will find that std::cout is not ready to write to (the stream is in an errored state); this seems to be due to a combination of an explicit call to close these streams and also calls that attempt to ask the "size" of streams such as stdout (cout). 

This pull request:

- adds a new interface called RunEnergyPlus(...) which preserves the previous interface's name (EnergyPlusPgm) and behavior (to exit via std::exit). This allows calling an energyplus simulation as a library without exiting the caller's process.
- proposes one possible way to handle stdin/stdout/stderr so as to resolve any error states upon return from the simulation call
- modifies `src/EnergyPlus/test_ep_as_library.cc` to demonstrate this newly working functionality

NOTE: The pull request ["Provide a C-API for the EnergyPlus DLL for ease of calling from other programming languages (6893)"](https://github.com/NREL/EnergyPlus/pull/6893) is a superset of this pull request. It includes all the changes suggested here and additionally adds a C-API for ease of calling the EnergyPlus DLL from other programming languages.

**Work completed under the Big Ladder Contribution License Agreement.**

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

michael-okeefe committed:
Add test C API for use with DLL

The HelloWorld function takes a string buffer (allocated on the client side) and the buffer
size and copies the string "Hello, World!" into the string buffer provided. This is a minimal
example of interaction with a DLL library meant to prove the connection.

michael-okeefe committed:
Explose API via extern "C"

HelloWorld and CheckConnect have been successfully accessed from Python, Delphi, and Ruby.
RunEPlus has been successfully demonstrated from Python and Ruby/Delphi to come.

michael-okeefe committed:
Prevent closing stdin, stdout

michael-okeefe committed:
Prevent queries that could put std::cout in an error state or close it

tellp returns -1 if it is not appropriate to get the position of the output stream (as for example for stdout/std::cout). Check explicitly if we have cout or if we return -1 from tellp(). Returns 0 for size in that case. Should it return -1 instead?

michael-okeefe committed:
Prevent library routine from exiting the process early.

Other calls to this function as a library are disrupted with this early exit.

michael-okeefe committed:
Provide test case that demonstrates call to EnergyPlus with a working cout and continuation of program after the call.

Prior to previous fixes, program would exit at EnergyPlusPgm and even if it didn't exit it would return with cout closed or in an error state (unusable). This demonstrates that the corrections have been applied correctly.

michael-okeefe committed:
Add new call; split out calling convention explicitly

michael-okeefe committed:
Protect against rewinding or seeking against streams such as stdin and stdout

michael-okeefe committed:
Add progress callback handler

michael-okeefe committed:
Return error code if we get in the catch block

Note: we should never enter the catch block but just in case...

michael-okeefe committed:
Remove old commented out lines

michael-okeefe committed:
Rename file

michael-okeefe committed:
Update CMake lists to add capi.{cc, hh} to build for energyplusapi. Change include to correspond to new header extension (*.hh)

michael-okeefe committed:
Create and expose a new function that uses return codes instead of hard exits

Calling exit(...) in library code is problematic as it doesn't allow the calling application to handle errors gracefully or provide error
recovery and/or cleanup. This change exposes a new return-code based library API function that can be linked to code to use
error codes similar to the exit status codes but without a hard application exit.

- the old functionality of hard exit is preserved under the original function name that uses the return code version under the hood
- prevent AbortEnergyPlus() from calling exit(...)
- keep old behavior to explicitly exit from EnergyPlusPgm for those depending on that behavior

michael-okeefe committed:
Remove CAPI files for pull request

michael-okeefe committed:
Updates to adjust style of code

- also allow exit code from EnergyPlusReturnCodes to come through naturally if not success code

michael-okeefe committed:
Merge branch 'return_codes' into return_codes_pull_request

# Conflicts:
#	src/EnergyPlus/EnergyPlusPgm.cc
#	src/EnergyPlus/UtilityRoutines.cc
#	src/EnergyPlus/public/EnergyPlusPgm.hh
#	src/EnergyPlus/test_ep_as_library.cc

michael-okeefe committed:
Simplify EnergyPlusPgm body

michael-okeefe committed:
Change calls to exit to return integer codes instead

michael-okeefe committed:
Explicitly namespace remaining (valid) calls to exit

exit(...) => std::exit(...)

Note: it would be ideal to eliminate all calls to exit() in the current program but many of the remaining exit(...) calls are for truly exceptional circumstances and/or are used by the command-line interface which still uses exit(...) by design.

michael-okeefe committed:
Change AbortEnergyPlus and EndEnergyPlus to explicitly return success/failure codes

michael-okeefe committed:
Revert changes to ObjexxFCL library

Leave changes to ObjexxFCL to upstream. Correct issue with stdout closing in EnergyPlus

michael-okeefe committed:
Add test that demonstrates fix for return codes and for stream issues.

michael-okeefe committed:
Clear std::cin and std::cout from being put into an error state

Standard input and standard output are placed into the ObjexxFCL global streams. Apparently, during iteration over these files via the FindUnitNumber(...) function in UtilityRoutines.cc, calls to gio::inquire(Unit cosnt unit, IOFlags &flags) which calls stream->size() cause these streams to be placed into an errored state.
It may be better that standard in and standard out are removed from the global streams listing. However, ObjexxFCL has been notified of the issue and they are considering changes for the next ObjexxFCL release that will fix the problem.

In the meantime, an easy way to handle the situation is to check if cin (standard in) or cout (standard out) are in an errored stae (i.e., not good) and clear their error flags if so.

The routine has been updated to clear cin/cout error flags if needed and to also print out to both cerr and cout to demonstrate cout is working.

michael-okeefe committed:
Prevent closing of standard error should it ever show up in the global streams

Also change how we query if we have a "special stream" (i.e., standard input, standard output, or standard error)

michael-okeefe committed:
Also clear any error state for the standard error stream

Myoldmopar commented on src/EnergyPlus/EnergyPlusPgm.cc:
```diff
@@ -224,6 +224,11 @@
+    std::exit(EnergyPlusPgmReturnCodes(filepath));
+}
+
+int EnergyPlusPgmReturnCodes(std::string const & filepath)
```
I really thought from the PR description that this function was going to give me the possible return codes from E+.  We'll figure out a better name.

jasondegraw replied:
Seems like `RunEnergyPlus` is the consensus on a better name.

Myoldmopar commented on src/EnergyPlus/UtilityRoutines.cc:
```diff
@@ -597,7 +597,7 @@ void AbortEnergyPlus()
     // indicating that E+ terminated with an error.
     if (NumExternalInterfaces > 0) CloseSocket(-1);
-    std::exit(EXIT_FAILURE);
+    return EXIT_FAILURE;
```
I was a bit nervous about this, because of course a call to ShowFatal needs to result in the program stopping instantly, not returning and accidentally continuing.  After some browsing in the code, however, I see that this is only called in one spot, EnergyPlusPgm, and only if an exception is thrown during the simulation, and is immediately followed by the EndEnergyPlus call.  Since ShowFatal raises an exception now (instead of calling AbortEnergyPlus), this is a reasonable approach.  :+1:

Myoldmopar commented on src/EnergyPlus/UtilityRoutines.cc:
```diff
@@ -719,12 +724,18 @@ void CloseOutOpenFiles()
             ios = flags.ios();
+            name = flags.name();
+        }
+        if (exists && opened && ios == 0) {
```
A little hacky, but makes sense.

jasondegraw commented on src/EnergyPlus/test_ep_as_library.cc:
```diff
@@ -64,10 +64,17 @@ int main(int argc, char *argv[])
+    if (!std::cin.good()) std::cin.clear();
+    if (!std::cerr.good()) std::cerr.clear();
+    if (!std::cout.good()) std::cout.clear();
+    std::cerr << "THIS LINE SHOULD DISPLAY." << std::endl;
```
Do we still need these lines here (this one and the next one)? If we do, I suggest changing them to print out "Standard error is still available for use" and "Standard output is still available for use". The context for the addition of these lines are here will fade over time, so it's better to change them to something more self explanatory.

michael-okeefe replied:
I like the suggested changes to the lines. @DeadParrot gave me a preview of the new ObjexxFCL and that did indeed fix the stdin/stdout issue but it may be nice to have code in this example use file that explicitly demonstrates that there is activity after the RunEnergyPlus call (showing that we're not just exiting anymore). That said, happy to go either way on whether we keep the lines or not.

jasondegraw reviewed (changes requested):
After going through the modifications, I'd like to see a couple changes: (1) change the function name (RunEnergyPlus seems to be the consensus) and (2) edit the messages at the end of the library test program. Other than those two minor things, it looks good to me.

michael-okeefe committed:
Update pull request per feedback

- Change EnergyPlusPgmReturnCodes ==> RunEnergyPlus
- Update example lines in test_ep_as_library.cc to print out "Standard error is still available for use" and "Standard output is still available for use".

nrel-bot-3 commented:
@michael-okeefe @lgentile it has been 14 days since this pull request was last updated.
